### PR TITLE
fix(core): vertex schema fallback, mcp log dedup, safe serialization, timeout handling

### PR DESCRIPTION
**Pull request by pdogra1299**
# Pull Request

## Description

**What does this PR do?**

Production-hardening fixes driven by a curator (consumer) log audit on OpenObserve, where these SDK behaviors showed up as recurring error signatures over the last 7 days.

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)

## Motivation and Context

A 7-day production log audit of the curator service surfaced five SDK-side failure classes:

- Vertex Gemini rejects complex response schemas with a deterministic 400 ("schema produces a constraint that has too many states for serving") and the call hard-failed (~850 consumer failures/week).
- One failed external MCP server registration emitted 5 ERROR log lines through the layers (~1,700 duplicate records/week).
- Eager `JSON.stringify` of full options/history in log metadata threw `RangeError: Invalid string length` on huge conversations and aborted the turn.
- `options.timeout` is applied per SDK step in the native Vertex loops, but a step TimeoutError was retried through the whole fallback chain with the same doomed budget (one 90s timeout became ~650s of retries).
- Vertex 429s surfaced as bare `ApiError status 429` with no model/region/retry context; Anthropic-on-Vertex 529 `overloaded_error` fell into the generic 5xx branch.

## Changes Made

- `structuredOutputPolicy.ts`: new `isSchemaComplexityError()`; `GenerationHandler` retries without structured output on that 400 (same flow as `isToolsSchemaConflictError`), and the native Gemini generate path retries once in schema-less JSON mode (`responseMimeType` kept, `responseSchema` dropped) — `generate({schema})` still returns coerced `structuredData`.
- MCP registration failures now log ONE error at the outermost owner (`NeuroLink.addExternalMCPServer`, enriched with transport/command/redacted url); inner layers log at debug; `.mcp-config.json` loader keeps its own per-server ERROR (it bypasses the public entry point); remove-after-failed-add "not found" is a debug no-op.
- New `safeDebugSerialize()` (bounded, never-throws) + `logger.shouldLog("debug")` gates for options dumps; `stringifyContentSafe()` for message/tool content on functional paths (`neurolink.ts`, `conversationMemory.ts`, `structuredSummarizer.ts`, native Anthropic history builders).
- Per-step timeout semantics documented on `TextGenerationOptions.timeout`; MCP retry loop surfaces a caller-set step TimeoutError immediately (auto-scaled internal timeouts still retry).
- Vertex rate-limit errors now report model + effective region (Gemini routes to `global`), scrape `retryDelay` from the error body, and classify 529/overloaded as rate-limit.

## Breaking Changes

- [x] No breaking changes

## Testing

- [x] Unit tests added/updated — `test/continuous-test-suite-json.ts` covers `isSchemaComplexityError` (21/21 pass)
- [x] Existing tests pass — `continuous-test-suite-mcp-infra.ts` 84/84
- [x] `tsc --noEmit --strict` exit 0; `pnpm run lint` 0 errors
- Changes were additionally reviewed by a multi-agent adversarial review pass (per-area reviewers + 3-refuter verification per finding).

## Commit Message Format

- [x] Commit message follows format: `type(scope): description`

## Dependencies

- [x] No dependency changes

## Summary by CodeRabbit

* **New Features**
  * Enhanced structured-output recovery: added a new “schema too complex” failure mode with targeted fallback/retry behavior.
  * More resilient processing of prompts, tool inputs/outputs, and conversation history to handle unusual/large payloads safely.

* **Bug Fixes**
  * Prevented token estimation and debug logging from crashing on extremely large or unserializable content.
  * Reduced duplicate or overly noisy logging during external server and client setup; adjusted log levels for expected failures.
  * Improved timeout enforcement so explicitly set step limits halt further retry escalation.

* **Documentation / Tests**
  * Clarified per-step timeout behavior and added deterministic coverage for schema-complexity classification.

### CODE REVIEW

File: src/lib/mcp/externalServerManager.ts
Comment (Tara-ag) [inline src/lib/mcp/externalServerManager.ts:1]: 💡 **Suggestion: Log level inconsistency in server loading error paths**

In the Promise.allSettled branch (lines ~486-496), errors are logged at `error` level for config-loaded servers. However, in the sequential loading branch (lines ~642-665), the same pattern applies but the catch block at line ~661 still logs at `error` level even though the error will be rethrown and handled at a higher layer.

Consider whether the catch block at line ~661 should also be `debug` level for consistency with the `addServer` method pattern (lines ~938-943), since the error propagates up and gets logged at the entry point.

Current:
```typescript
} catch (error) {
  // See above: outermost point for config-loaded servers.
  mcpLogger.error(`[ExternalServerManager] ${errorMsg}`);
}
```

Suggested (if the error is rethrown/handled elsewhere):
```typescript
} catch (error) {
  // debug: error propagates to outermost handler
  mcpLogger.debug(`[ExternalServerManager] ${errorMsg}`);
}
```

File: src/lib/providers/googleVertex.ts
Comment (Tara-ag) [inline src/lib/providers/googleVertex.ts:1]: 💡 **Suggestion: Code duplication of safe stringify function**

The `stringifyHistoryContentSafe` function (lines 488-502) is nearly identical to `stringifyContentSafe` in `neurolink.ts` (lines 324-338). Consider exporting the function from a shared utility module to avoid duplication and ensure consistent behavior.

Current in googleVertex.ts:
```typescript
const stringifyHistoryContentSafe = (content: unknown): string => {
  if (typeof content === "string") {
    return content;
  }
  try {
    return JSON.stringify(content) ?? String(content);
  } catch {
    return "[content too large to serialize]";
  }
};
```

Same pattern exists in neurolink.ts. Consider moving to `src/lib/utils/logSanitize.ts` alongside `safeDebugSerialize`.

File: test/continuous-test-suite-json.ts
Comment (Tara-ag) [inline test/continuous-test-suite-json.ts:1]: 💡 **Suggestion: Add test case for null input to isSchemaComplexityError**

The test suite covers `undefined` input but doesn't explicitly test `null` input. Consider adding a test case to ensure consistent behavior:

```typescript
assertEqual(
  isSchemaComplexityError(null),
  false,
  "null must not match",
);
```

This ensures the function handles all falsy values consistently.

**Review by Tara-ag — COMMENTED** [review]
## Review Summary

This PR addresses five production-hardening issues identified through a 7-day log audit. Overall, the changes are well-structured and follow the project's patterns.

### Files Reviewed: 11
- `src/lib/context/stages/structuredSummarizer.ts` ✅
- `src/lib/core/modules/GenerationHandler.ts` ✅
- `src/lib/core/modules/structuredOutputPolicy.ts` ✅
- `src/lib/mcp/externalServerManager.ts` 💡
- `src/lib/mcp/mcpClientFactory.ts` ✅
- `src/lib/neurolink.ts` 💡
- `src/lib/providers/googleVertex.ts` 💡
- `src/lib/types/generate.ts` ✅
- `src/lib/utils/conversationMemory.ts` ✅
- `src/lib/utils/logSanitize.ts` ✅
- `test/continuous-test-suite-json.ts` 💡

### Issues Found: 3 (all MINOR/SUGGESTION)

| Severity | Count | Description |
|----------|-------|-------------|
| 💡 SUGGESTION | 3 | Code duplication, test coverage, log level consistency |

### Key Observations

**✅ Well-implemented:**
- Safe serialization with `safeDebugSerialize()` prevents `RangeError` on large objects
- `isSchemaComplexityError()` correctly detects Vertex "too many states" errors
- Timeout handling properly distinguishes caller-set vs auto-scaled timeouts
- MCP log deduplication strategy is sound (single ERROR at entry point, debug in inner layers)
- Vertex 529/overloaded errors now correctly classified as rate-limit

**💡 Suggestions made:**
1. **Code duplication**: `stringifyContentSafe` and `stringifyHistoryContentSafe` are nearly identical - consider sharing via `logSanitize.ts`
2. **Test coverage**: Added suggestion for `null` input test case for `isSchemaComplexityError`
3. **Log level consistency**: One catch block in externalServerManager.ts could use debug level for consistency

### Security Check
- ✅ No hardcoded secrets detected
- ✅ URL credentials properly redacted via `redactUrlCredentials()`
- ✅ Safe serialization prevents information leakage in logs

### Backward Compatibility
- ✅ No breaking changes to public SDK API
- ✅ New timeout semantics are documented in `GenerateOptions.timeout` JSDoc

### Verdict
**Approve with suggestions.** The PR addresses critical production issues effectively. The suggestions are minor improvements that can be addressed in follow-up work or accepted as-is.

File: src/lib/mcp/externalServerManager.ts
Comment (pdogra1299) [reply src/lib/mcp/externalServerManager.ts:1]: Not applied — this one is a false positive: unlike `addServer`/`startServer` (which rethrow/return into `NeuroLink.addExternalMCPServer`, the entry point that owns the ERROR), this sequential-load catch does **not** rethrow — it `continue`s with the remaining servers, and config-loaded servers never pass through `addExternalMCPServer` at all (`loadMCPConfiguration` calls `this.addServer` directly). So this catch *is* the outermost point for the config path; downgrading it to debug would leave `.mcp-config.json` startup failures with zero ERROR records — exactly the regression flagged and fixed for the parallel path in this PR.

File: src/lib/mcp/externalServerManager.ts [hunk 1]
```diff
@@ -486,10 +486,17 @@ export class ExternalServerManager extends EventEmitter {
...
+            mcpLogger.error(
+              `[ExternalServerManager] Failed to load server ${serverId}: ${error}`,
+            );
           } else if (serverResult && !serverResult.success) {
             const errorMsg = `Failed to load server ${serverId}: ${serverResult.error}`;
             errors.push(errorMsg);
-            mcpLogger.warn(`[ExternalServerManager] ${errorMsg}`);
+            // See above: outermost point for config-loaded servers.
+            mcpLogger.error(`[ExternalServerManager] ${errorMsg}`);
           }
```
[the remark this answers is not in the document]
Comment (pdogra1299) [reply src/lib/mcp/externalServerManager.ts:502]: Fixed in 814b8af4 — removed the inner task-catch `mcpLogger.warn`; the result-processing loop below now owns the single record per failed parallel-load server.

File: src/lib/neurolink.ts [hunk 2]
```diff
@@ -9813,14 +9851,20 @@ Current user's request: ${currentInput}`;
...
+    if (logger.shouldLog("debug")) {
+      logger.debug(
+        "[NeuroLink.stream] Preparing to store conversation turn in memory",
+        {
+          options: safeDebugSerialize(enhancedOptions),
+          sessionId: (enhancedOptions.context as Record<string, unknown>)
+            ?.sessionId,
+        },
+      );
+    }
```
[the remark this answers is not in the document]
Comment (pdogra1299) [reply src/lib/neurolink.ts:9853]: Fixed in 814b8af4 — both dumps (plus the two in `conversationMemory.ts`) now run `sanitizeRecord()` (deep, key-based redaction + per-string truncation + cycle guard) before `safeDebugSerialize()`, and the sensitive-key list gained `authContext`/`authToken`/`sessionToken`/`serviceAccountKey`/`secretAccessKey`/`userEmail` (`credentials` was already covered). Also applied your outside-diff catch: the warn-path `options` field in `conversationMemory.ts` is now gated behind `logger.shouldLog("debug")`.

File: src/lib/providers/googleVertex.ts [hunk 3]
```diff
@@ -487,6 +488,23 @@ export const resolveVertexLocation = (
...
+  if (typeof content === "string") {
+    return content;
+  }
+  try {
+    return JSON.stringify(content) ?? String(content);
+  } catch {
+    return "[content too large to serialize]";
+  }
+};
+
```
[the remark this answers is not in the document]
Comment (pdogra1299) [reply src/lib/providers/googleVertex.ts]: Fixed in 814b8af4 — both live Anthropic tool-result sites (stream + generate loops) now use the shared `stringifyContentSafe()`, so a serialization failure after a successful tool call can no longer be misreported as a tool execution failure.

File: src/lib/providers/googleVertex.ts [hunk 4]
```diff
@@ -6126,8 +6145,46 @@ export class GoogleVertexProvider extends BaseProvider {
...
+                          nativeError instanceof Error
+                            ? nativeError.message
+                            : String(nativeError),
+                      },
+                    );
+                    nativeResult = await this.executeNativeGemini3Generate({
+                      ...mergedOptions,
+                      schema: undefined,
+                      output: { format: "json" },
+                    });
```
[the remark this answers is not in the document]
Comment (pdogra1299) [reply src/lib/providers/googleVertex.ts:6197]: Fixed in 814b8af4 — added the `hasTools` guard as suggested: with tools present the retry is skipped and the 400 surfaces (the AI-SDK path keeps its own GenerationHandler text-mode fallback); only the no-tools case retries, where `output.format: "json"` genuinely keeps `responseMimeType: application/json`.

File: src/lib/providers/googleVertex.ts
Comment (pdogra1299) [reply src/lib/providers/googleVertex.ts:1]: Fixed in 814b8af4 — `stringifyContentSafe` now lives in `src/lib/utils/logSanitize.ts` next to `safeDebugSerialize`; both local copies (neurolink.ts + the googleVertex duplicate) are removed and import the shared export.

File: test/continuous-test-suite-json.ts
Comment (pdogra1299) [reply test/continuous-test-suite-json.ts:1]: Fixed in 814b8af4 — added the `null` case to the detector test (json suite 21/21).

**Review by Tara-ag — APPROVED** [review]
## Review Summary

This PR successfully addresses **5 production-hardening issues** identified through log auditing:

### Changes Verified ✅

1. **Vertex Gemini schema complexity handling** - New `isSchemaComplexityError()` detector with graceful fallback to schema-less JSON mode while preserving `generate({schema})` contract

2. **MCP log deduplication** - Inner layers now log at `debug` level; single ERROR record at `NeuroLink.addExternalMCPServer` entry point with enriched context (transport, command, redacted URL)

3. **Safe serialization** - New `safeDebugSerialize()` and `stringifyContentSafe()` utilities prevent `RangeError: Invalid string length` on large conversations; all call sites gated with `logger.shouldLog("debug")`

4. **Timeout handling** - Caller-set `options.timeout` captured before auto-scaling; step TimeoutError surfaces immediately instead of burning retries

5. **Enhanced error classification** - Vertex 529/overloaded errors now classified as rate-limit with model/region context and retry delay scraping

### Code Quality
- All existing review comments addressed (814b8af4)
- Tests added for `isSchemaComplexityError` (21/21 pass)
- No hardcoded secrets
- `formatProviderError` correctly returns (never throws)
- Logger guards properly applied

### Minor Suggestions (non-blocking)
1. Consider tightening the `isSchemaComplexityError` regex to `/schema produces? a constraint that has too many states/i` for precision
2. Document `stringifyContentSafe` behavior for `undefined` inputs
3. Consider adding span attribute for schema complexity fallback metrics

**Approved for merge.**

File: src/lib/providers/googleVertex.ts [hunk 5]
```diff
@@ -6126,8 +6129,55 @@ export class GoogleVertexProvider extends BaseProvider {
...
```
[the remark this answers is not in the document]
Comment (pdogra1299) [reply src/lib/providers/googleVertex.ts:6197]: Not applied — verified false positive: `structuredData` coercion does not happen in the provider layer at all. Neither `googleVertex.ts` nor `BaseProvider.enhanceResult` reads `options.schema` for coercion (zero `structuredData` references in either); the coercion runs at the NeuroLink layer in the generate finalization path (`neurolink.ts` ~4791: `coerceJsonToSchema(textResult.content, textOptions.schema)`), which uses the CALLER's `textOptions` — untouched by the provider-internal `mergedOptions` clone. The schema-less retry returns JSON-mode text, and the layer above still coerces it against the original schema, so the `generate({schema})` guarantee holds. Also covered by the `isSchemaComplexityError` regression test + json suite (21/21).

File: src/lib/providers/googleVertex.ts [hunk 6]
```diff
@@ -6601,20 +6651,43 @@ export class GoogleVertexProvider extends BaseProvider {
...
+        this.location,
+      );
       return new RateLimitError(
-        `Google Vertex AI quota/rate limit exceeded. ` +
-          `Solutions: 1. Check your Vertex AI quotas in Google Cloud Console ` +
-          `2. Request quota increase if needed ` +
-          `3. Try a different model or reduce request frequency ` +
-          `4. Consider using a different region`,
+        `Google Vertex AI rate limit / shared-capacity exhausted (429 RESOURCE_EXHAUSTED / overloaded) ` +
+          `for model '${this.modelName}' in region '${effectiveRegion}'.` +
```
[the remark this answers is not in the document]
Comment (pdogra1299) [reply src/lib/providers/googleVertex.ts]: Fixed in d8464f62 — the two native catch sites now attach best-effort `requestModel` (and `requestRegion` on the Gemini path) to the error before `handleProviderError`, and the rate-limit branch in `formatProviderError` prefers those over `this.modelName`/`this.location`, falling back to the instance values for non-native paths.

File: src/lib/utils/logSanitize.ts [hunk 7]
```diff
@@ -100,6 +107,55 @@ export function sanitizeForLog(text: string, maxLen = 500): string {
...
+    return content;
+  }
+  try {
+    return JSON.stringify(content) ?? String(content);
+  } catch {
+    return "[content too large to serialize]";
+  }
+}
+
+export function safeDebugSerialize(value: unknown, maxLen = 10_000): string {
```
[the remark this answers is not in the document]
Comment (pdogra1299) [reply src/lib/utils/logSanitize.ts]: Fixed in d8464f62 — good catch: the insertion had landed between `safeDebugSerialize`'s JSDoc and its function. Each doc block now sits directly above its own function.

**Review by Tara-ag — APPROVED** [review]
## Review Summary

I've completed a comprehensive review of PR #1126 covering all 11 changed files.

### ✅ Verification Complete

**CLAUDE.md Critical Rules:**
- Rule 6 (`formatProviderError` returns, never throws): VERIFIED - All error paths return typed errors
- Rule 1 (dynamic imports only in registry): VERIFIED - No static provider imports outside registry
- Rule 5 (backward compatibility): VERIFIED - No breaking changes to public SDK API

**Security:**
- No hardcoded secrets or credentials
- URL credentials properly redacted via `redactUrlCredentials()`
- Sensitive keys expanded in `SENSITIVE_OBJECT_KEYS`
- Safe serialization prevents PII leakage in debug logs

**Production Hardening (5 issues addressed):**
1. ✅ Vertex schema complexity fallback with `isSchemaComplexityError()`
2. ✅ MCP log deduplication (single ERROR at entry point)
3. ✅ Safe serialization prevents `RangeError: Invalid string length`
4. ✅ Caller-set timeout handling (surfaces immediately, no doomed retries)
5. ✅ Enhanced Vertex error classification (529/overloaded as rate-limit)

**Code Quality:**
- Logger guards properly applied with `logger.shouldLog("debug")`
- Tests added for schema complexity detection (21/21 pass)
- Shared utilities prevent code duplication
- Proper error context for observability

### Prior Review Resolution
All comments from previous reviewers (Tara-ag, CodeRabbit) have been addressed in commits 814b8af4 and d8464f62.

**APPROVED for merge.**
